### PR TITLE
Reposync bugfixes

### PR DESCRIPTION
**Pull request by moio**
## What does this PR change?

Bugfixes to regressions introduced by reposync speedups: #2154 #2157 #2162 #2164 

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: **bugfixes**

- [x] **DONE**

## Test coverage
- No tests: **will be covered in upcoming integration tests**

- [x] **DONE**

## Links

See also #2154 #2157 #2162 #2164 

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by moio** [comment]
Thanks @mcalmer for reporting, this should fix both problems you observed.

:man_facepalming: about the first one...!!!

**Comment by moio** [comment]
@mcalmer what channel are you synchronizing so that I can try to reproduce here?

**Comment by mcalmer** [comment]
@moio res7-x86_64 - I wanted to know how long it will take.

I used a filter to exclude all "debuginfo" packages, because they would not fit on the disk.

```
$> mgr-sync add product --no-sync
... select RHEL7 Expanded Support product
$> spacecmd api -- channel.software.addVendorRepoFilter -A '["res7-x86_64", {"filter": "*-debuginfo", "flag": "-"}]'
$> mgr-sync sync channels --with-children
... select the base channel for RH7
```

**Comment by juliogonzalez** [comment]
I can reproduce this with Ubuntu 20.04 from https://github.com/SUSE/spacewalk/issues/11367

And I had the same when I used the 18.04 by mistake.

This is from `ubuntu-2004-amd64-main-security-uyuni.log`:
```
2020/05/04 16:49:46 +02:00     202/204 : mysql-server-core-8.0_8.0.20-0ubuntu0.20.04.1_amd64.deb
2020/05/04 16:49:46 +02:00     203/204 : samba-libs_4.11.6+dfsg-0ubuntu1.1_amd64.deb
2020/05/04 16:49:50 +02:00     204/204 : linux-source-5.4.0_5.4.0-28.32_all.deb
2020/05/04 16:49:51 +02:00 Importing packages started.
2020/05/04 16:49:51 +02:00 
2020/05/04 16:49:51 +02:00   Importing packages to DB:
2020/05/04 16:49:52 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:52 +02:00   Package batch #2 of 11 completed...
2020/05/04 16:49:53 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:53 +02:00   Package batch #4 of 11 completed...
2020/05/04 16:49:53 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:53 +02:00   Package batch #3 of 11 completed...
2020/05/04 16:49:53 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:53 +02:00   Package batch #8 of 11 completed...
2020/05/04 16:49:53 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:53 +02:00   Package batch #6 of 11 completed...
2020/05/04 16:49:53 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:53 +02:00   Package batch #1 of 11 completed...
2020/05/04 16:49:53 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:53 +02:00   Package batch #11 of 11 completed...
2020/05/04 16:49:54 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:54 +02:00   Package batch #10 of 11 completed...
2020/05/04 16:49:54 +02:00   Package batch #7 of 11 completed...
2020/05/04 16:49:54 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:54 +02:00   Package batch #9 of 11 completed...
2020/05/04 16:49:54 +02:00 '<' not supported between instances of 'str' and 'NoneType'
2020/05/04 16:49:54 +02:00   Package batch #5 of 11 completed...
2020/05/04 16:49:55 +02:00 Importing packages finished.
2020/05/04 16:49:55 +02:00 
2020/05/04 16:49:55 +02:00   Linking packages to the channel.
2020/05/04 16:49:55 +02:00 Unexpected error: <class 'TypeError'>
2020/05/04 16:49:55 +02:00 Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 590, in sync
    ret = self.import_packages(plugin, data['id'], url, is_non_local_repo)
  File "/usr/lib/python3.6/site-packages/spacewalk/satellite_tools/reposync.py", line 1096, in import_packages
    importer.run()
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/importLib.py", line 774, in run
    self.fix()
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/packageImport.py", line 76, in fix
    self.backend.lookupEVRs(self.evrs)
  File "/usr/lib/python3.6/site-packages/spacewalk/server/importlib/backend.py", line 671, in lookupEVRs
    for evr in sorted(evrHash.keys(), key=lambda x: (x[0] if x[0] != '' else None, x[1], x[2])):
TypeError: '<' not supported between instances of 'str' and 'NoneType'

**Comment by moio** [comment]
I just force-pushed a fix that corrects the problem @juliogonzalez reported and possibly also [the last one from @mcalmer](https://github.com/uyuni-project/uyuni/pull/2181#pullrequestreview-404616080).

I still cannot say for sure as I am still not done fixing up my RES 7 environment - I hope to be back with certain news ASAP.

Meanwhile it'd be great if you could re-test.

**Comment by juliogonzalez** [comment]
ACK. Will give it a try tomorrow morning and will let you know :-)

File: backend/server/importlib/backend.py [hunk 1]
```diff
@@ -668,7 +669,7 @@ def lookupErrataSeverityId(self, erratum):
     def lookupEVRs(self, evrHash):
         sql = "select LOOKUP_EVR(:epoch, :version, :release) id from dual"
         h = self.dbmodule.prepare(sql)
-        for evr in sorted(evrHash.keys(), key=lambda x: (x[0] if x[0] != '' else None, x[1], x[2])):
+        for evr in sorted(evrHash.keys(), key=lambda k: (str(k or 0), k[1], k[2])):
```
Comment (mcalmer) [inline backend/server/importlib/backend.py:672]: (k or 0) sounds wrong. Should it be `(k[0] or 0)` ?

**Comment by juliogonzalez** [comment]
Looks the import works now: 
image

I can't see anymore stacktraces on the logs.

I will test onboarding soon, to verify that nothing else is broken.

File: backend/server/importlib/backend.py (hunk 1 above)
Comment (moio) [reply backend/server/importlib/backend.py]: Right, addressed

**Comment by mcalmer** [comment]
@moio JFYI: I tried this on refhead with res7 (excluded debuginfo packages) on 2 CPUs only.
It took 3h 30m . Tomorrow I want to try with 6CPUs.

**Comment by moio** [comment]
I am also testing with 6CPUs here today, and will hopefully have results tomorrow morning.

Unless problems arise I will merge tomorrow early afternoon.

**Comment by mcalmer** [comment]
@moio My test with 6 CPUs was finished after 2h 10m. Time to merge.

**Comment by moio** [comment]
Fore the record in my setup with 6 CPUs it took 2:47 with all packages, including debuginfos

**Pull request merged**